### PR TITLE
[RFC] [3.3] Add support for associations in custom datasource

### PR DESCRIPTION
**Pull request by Marlinc**
These changes make it possible to use associations in custom datasources. It also moves some logic that is not very database ORM specific to the datasource component.

This is still a work in progress and comments are appreciated!

### CODE REVIEW

File: src/ORM/AssociationCollection.php [hunk 1]
```diff
@@ -15,6 +15,7 @@
 namespace Cake\ORM;
 
 use ArrayIterator;
+use Cake\Datasource\AssociationsNormalizerTrait;
```
Comment (markstory) [inline src/ORM/AssociationCollection.php]: Doesn't this leave us with a lot of duplicated code between ORM and Datasour e packages?

**Comment by lorenzo** [comment]
Is thee an alternative implementation of associations already available? I fear that we may be abstracting something without thinking if this is the right abstraction

**Comment by Marlinc** [comment]
I'm currently working on one for the Webservices plugin. I opened the PR so that the interface could be discussed :smile:

File: src/Datasource/EagerLoaderInterface.php [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+ */
+namespace Cake\Datasource;
+
+/**
+ * Exposes the methods for storing the associations that should be eager loaded
+ * for a repository once a query is provided and delegates the job of creating the
+ * required joins and decorating the results so that those associations can be
+ * part of the result set.
+ */
+interface EagerLoaderInterface
```
Comment (lorenzo) [inline src/Datasource/EagerLoaderInterface.php]: A good start would be not having this as an interface. This class is pretty much knowledgable of how joins work in SQL.

File: src/Datasource/AssociationInterface.php [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+     * - sort: The direction in which the records should be returned
+     * - fields: List of fields to select from the target repository
+     * - contain: List of related repositories to eager load associated to the target repository
+     * - strategy: The name of strategy to use for finding target repository records
+     * - nestKey: The array key under which results will be found when transforming the row
+     *
+     * @param array $options The options for eager loading.
+     * @return \Closure
+     */
+    public function eagerLoader(array $options);
```
Comment (lorenzo) [inline src/Datasource/AssociationInterface.php]: I don't think this function should be part of the interface

File: src/Datasource/AssociationInterface.php [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+     * - negateMatch: Will append a condition to the passed query for excluding matches.
+     *   with this association.
+     *
+     * @param \Cake\Datasource\QueryInterface $query the query to be altered to include the target repository data
+     * @param array $options Any extra options or overrides to be taken in account
+     * @return void
+     * @throws \RuntimeException if the query builder passed does not return a query
+     * object
+     */
+    public function attachTo(QueryInterface $query, array $options = []);
```
Comment (lorenzo) [inline src/Datasource/AssociationInterface.php]: This function should probably also be left out of the interface

File: src/Datasource/AssociationInterface.php [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+     * Correctly nests a result row associated values into the correct array keys inside the
+     * source results.
+     *
+     * @param array $row The row to transform
+     * @param string $nestKey The array key under which the results for this association
+     *   should be found
+     * @param bool $joined Whether or not the row is a result of a direct join
+     *   with this association
+     * @return array
+     */
```
Comment (lorenzo) [inline src/Datasource/AssociationInterface.php]: Same with this one, not sure it has a lot of value inside an interface

**Comment by lorenzo** [comment]
I think most of it makes sense, except for the functions I pointed out

File: src/ORM/AssociationCollection.php (hunk 1 above)
Comment (Marlinc) [inline src/ORM/AssociationCollection.php]: I moved `AssociationsNormalizerTrait` to the datasource package because its not very specific to the ORM.

Comment (markstory) [inline src/ORM/AssociationCollection.php]: I was trying to comment on the AssociationCollection as a whole. The ORM version is still there.

**Comment by markstory** [comment]
Closing as this can't be merged. If you have time @Marlinc to get this into a mergable state, we can discuss it further.

File: src/ORM/AssociationCollection.php (hunk 1 above)
Comment (Marlinc) [inline src/ORM/AssociationCollection.php]: Thanks, I didn't see that. This has been resolved :+1:

**Comment by Marlinc** [comment]
I've rebased everything and resolved the things that were mentioned.

File: src/Datasource/EagerLoadable.php [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+     * - aliasPath
+     * - propertyPath
+     * - forMatching
+     *
+     * The keys maps to the set repository properties in this class.
+     *
+     * @param string $name The Association name.
+     * @param array $config The list of properties to set.
+     */
+    public function __construct($name, array $config = [])
```
Comment (ADmad) [inline src/Datasource/EagerLoadable.php]: Codecov says there's no unit test coverage for this class.

File: src/Datasource/AssociationCollection.php [hunk 7]
```diff
@@ -161,19 +161,18 @@ public function removeAll()
...
      * @param array $options The options for the save operation.
      * @return bool Success
      */
-    public function saveParents(Table $table, EntityInterface $entity, $associations, array $options = [])
+    public function saveParents(RepositoryInterface $table, EntityInterface $entity, $associations, array $options = [])
     {
         if (empty($associations)) {
             return true;
         }
-
```
Comment (ADmad) [inline src/Datasource/AssociationCollection.php]: New line needed here.

Comment (Marlinc) [inline src/Datasource/AssociationCollection.php]: This has now been fixed.

File: src/Datasource/EagerLoadable.php (hunk 6 above)
Comment (Marlinc) [inline src/Datasource/EagerLoadable.php]: Thank you, I'll look at this!

**Comment by josegonzalez** [comment]
This needs another rebase :(

**Comment by dereuromark** [comment]
Is this BC if you move classes without class_aliasing them?

    src/ORM/AssociationCollection.php → src/Datasource/AssociationCollection.php

etc

**Comment by markstory** [comment]
Closing as this can't be merged anymore. If it can be rebased we can discuss it further.

**Comment by ceeram** [comment]
This is gonna be picked up again, im reopening this PR. Reopening before rebasing is needed since there is a bug in github. See: https://github.com/isaacs/github/issues/361

**Comment by ceeram** [comment]
ping @Marlinc

**Comment by ceeram** [comment]
Closing again for the time being, rebase has been done, can be reopened later when @marlinc pushes more changes to further discuss

**Comment by ceeram** [comment]
reopening again as erquested by @Marlinc

**Comment by markstory** [comment]
Closing until this can be put into a state where it can be merged.

**Pull request closed without merging**